packages/flutter/lib/src/widgets/text.dart

```diff
@@ -1242,162 +1242,19 @@ class _SelectableTextContainerDelegate extends MultiSelectableSelectionContainer
+      lastStartEdgeUpdateGlobalPosition = event.globalPosition;
     }
-
-    return super.handleSelectionEdgeUpdate(event);
-  }
-
-  @override
-  void dispose() {
-    _hasReceivedStartEvent.clear();
-    _hasReceivedEndEvent.clear();
-    super.dispose();
-  }
-
-  @override
-  SelectionResult dispatchSelectionEventToChild(Selectable selectable, SelectionEvent event) {
-    switch (event.type) {
-      case SelectionEventType.startEdgeUpdate:
-        _hasReceivedStartEvent.add(selectable);
-        ensureChildUpdated(selectable);
-      case SelectionEventType.endEdgeUpdate:
-        _hasReceivedEndEvent.add(selectable);
-        ensureChildUpdated(selectable);
-      case SelectionEventType.clear:
-        _hasReceivedStartEvent.remove(selectable);
-        _hasReceivedEndEvent.remove(selectable);
-      case SelectionEventType.selectAll:
-      case SelectionEventType.selectWord:
-      case SelectionEventType.selectParagraph:
-        break;
-      case SelectionEventType.granularlyExtendSelection:
-      case SelectionEventType.directionallyExtendSelection:
-        _hasReceivedStartEvent.add(selectable);
-        _hasReceivedEndEvent.add(selectable);
-        ensureChildUpdated(selectable);
-    }
-    return super.dispatchSelectionEventToChild(selectable, event);
-  }
-
-  @override
-  void ensureChildUpdated(Selectable selectable) {
-    if (_lastEndEdgeUpdateGlobalPosition != null && _hasReceivedEndEvent.add(selectable)) {
-      final SelectionEdgeUpdateEvent synthesizedEvent = SelectionEdgeUpdateEvent.forEnd(
-        globalPosition: _lastEndEdgeUpdateGlobalPosition!,
-      );
-      if (currentSelectionEndIndex == -1) {
-        handleSelectionEdgeUpdate(synthesizedEvent);
-      }
-      selectable.dispatchSelectionEvent(synthesizedEvent);
-    }
-    if (_lastStartEdgeUpdateGlobalPosition != null && _hasReceivedStartEvent.add(selectable)) {
-      final SelectionEdgeUpdateEvent synthesizedEvent = SelectionEdgeUpdateEvent.forStart(
-          globalPosition: _lastStartEdgeUpdateGlobalPosition!,
-      );
-      if (currentSelectionStartIndex == -1) {
-        handleSelectionEdgeUpdate(synthesizedEvent);
-      }
-      selectable.dispatchSelectionEvent(synthesizedEvent);
-    }
-  }
-
-  @override
-  void didChangeSelectables() {
-    if (_lastEndEdgeUpdateGlobalPosition != null) {
-      handleSelectionEdgeUpdate(
-        SelectionEdgeUpdateEvent.forEnd(
-          globalPosition: _lastEndEdgeUpdateGlobalPosition!,
-        ),
-      );
-    }
-    if (_lastStartEdgeUpdateGlobalPosition != null) {
-      handleSelectionEdgeUpdate(
-        SelectionEdgeUpdateEvent.forStart(
-          globalPosition: _lastStartEdgeUpdateGlobalPosition!,
-        ),
-      );
+    if (event.type == SelectionEventType.endEdgeUpdate) {
+      return currentSelectionEndIndex == -1 ? _initSelection(event, isEnd: true) : _adjustSelection(event, isEnd: true);
```

Comment:
I think i missed this, why do we need to duplicate these two method in this class?

Reply:
I did this to handle the behavior where to keep the origin boundary within the selection we swap the selection edges when they invert. We didn't need this for word boundary because a word boundary is guaranteed to be within one `_SelectableFragment`, but with paragraphs the origin boundary may span multiple `_SelectableFragment`s so I needed to account for these swaps at the container level so it has an accurate `currentSelectionStartIndex` and `currentSelectionEndIndex`.